Add support for forwarding trace messages to logs

falk-haleytek opened the pull request:
Initially, support for tracing to DLT was added. Later, support for forwarding trace messages to logs on Android was added. This commit adds support for forwarding trace messages to logs on all platforms by introducing the build flag TRACE_TO_LOGS.

If either USE_DLT or TRACE_TO_LOGS is defined when building vsomeip, support for generating trace messages will be included in the library.

The Android.bp file has been updated to include TRACE_TO_LOGS instead of USE_DLT, which allows removal of a number of additional ifdef guards around DLT-specific code.

duartenfonseca commented:
After reviewing this PR with our internal team, we decided that the cost to benefit of this change, the testing and integration of this for the benefit of removing some of the ifdef guards would not pay off.

falk-haleytek commented:
@duartenfonseca Thank you for taking the time to review this PR.

The primary goal of this change is to extend support for outputting trace information to logs, beyond just DLT. While the main branch already allows trace output to logs on Android, this update enables similar functionality across other platforms where it can also be valuable.

The key addition in this PR can be seen here: [[link to the specific line in the PR]](https://github.com/COVESA/vsomeip/pull/726/files#diff-081b28badbad44c1580fc384ff2ccb684e2aab8b9931eb17f27e9fb8df7d9baeR280). 

This change introduces a TRACE_TO_LOGS directive, which decouples the logging functionality from the existing Android-specific ifdef. The intent is to clarify that tracing can be directed to both DLT (USE_DLT) and logs (TRACE_TO_LOGS) across all platforms, not just Android.

I would appreciate it if you could reconsider the decision to close this PR. If there are any adjustments or improvements you’d like to see to better align with the project's goals, I’m more than happy to make those changes to ensure that tracing to logs is supported on platforms beyond Android.

duartenfonseca commented:
@falk-haleytek i will take another look at this changes

duartenfonseca commented on implementation/tracing/src/connector_impl.cpp:
```diff
@@ -206,7 +198,7 @@ std::shared_ptr<channel_impl> connector_impl::get_channel_impl(const std::string
         const byte_t *_data, uint32_t _data_size) {
 
-#if USE_DLT
+#if defined(USE_DLT) || defined(TRACE_TO_LOGS)
```
```suggestion

```

duartenfonseca commented on implementation/tracing/src/connector_impl.cpp:
```diff
@@ -269,7 +261,8 @@ void connector_impl::trace(const byte_t *_header, uint16_t _header_size,
                     VSOMEIP_ERROR << "tracing: found channel without DLT context!";
                 }
-            #else
+            #endif
+            #ifdef TRACE_TO_LOGS
```
```suggestion
            #else
```

duartenfonseca replied:
```suggestion
            #else
```

duartenfonseca commented on implementation/tracing/src/connector_impl.cpp:
```diff
@@ -280,9 +273,12 @@ void connector_impl::trace(const byte_t *_header, uint16_t _header_size,
+                    std::string app = runtime::get_property("LogApplication");
+                    ALOGI(app.c_str(), ss.str().c_str());
+                #else
+                    VSOMEIP_INFO << ss.str();
```
```suggestion
                    VSOMEIP_TRACE << ss.str();
```

duartenfonseca commented:
@falk-haleytek would you agree with reverting the change of TRACE_TO_LOGS, which would be hard to change internally, but keep the added functionality of printing TC logs to the console? the suggestions I added would be for that purpose, but it would also mean that you would revert all the other TRACE_TO_LOGS changes

falk-haleytek commented:
Thanks for the new suggestions, @duartenfonseca.

Could you clarify what makes the TRACE_TO_LOGS change difficult to implement internally? Understanding this would help us find a solution that addresses your concerns.

Currently, the USE_DLT define seems to be somewhat misaligned with its intended use. It was [introduced](https://github.com/COVESA/vsomeip/commit/826ebb8d352245a36ecaec32b6af61e7abf4696e#diff-757cd72756ef920fd2357653c76c087315969557e257a0fc8532ed2c6dca1649R85) by default in the Android build, possibly to enable tracing to logs (?).  However, this is effectively negated by other additions, such as [this one](https://github.com/COVESA/vsomeip/commit/826ebb8d352245a36ecaec32b6af61e7abf4696e#diff-9289ab59b899e3cdadcc95b56fa9c04b3b34a672868accf6a5ed39d9dfeefe97R19), which guards against the USE_DLT define if it detects an Android build. This creates a situation where USE_DLT must be defined for Android, even though Android does not support DLT.

The introduction of a new define (TRACE_TO_LOGS) with a clear intent is meant to address this technical debt. It decouples the logging functionality from the Android-specific ifdef, making it clear that tracing can be directed to logs across all platforms, not just Android.

If there’s a way to retain the added functionality while addressing your internal concerns, I’m open to adjusting the implementation. I believe it’s important to resolve this in a way that benefits all platforms without introducing further confusion.

duartenfonseca commented:
> Thanks for the new suggestions, @duartenfonseca.
> 
> Could you clarify what makes the TRACE_TO_LOGS change difficult to implement internally? Understanding this would help us find a solution that addresses your concerns.
> 
> Currently, the USE_DLT define seems to be somewhat misaligned with its intended use. It was [introduced](https://github.com/COVESA/vsomeip/commit/826ebb8d352245a36ecaec32b6af61e7abf4696e#diff-757cd72756ef920fd2357653c76c087315969557e257a0fc8532ed2c6dca1649R85) by default in the Android build, possibly to enable tracing to logs (?). However, this is effectively negated by other additions, such as [this one](https://github.com/COVESA/vsomeip/commit/826ebb8d352245a36ecaec32b6af61e7abf4696e#diff-9289ab59b899e3cdadcc95b56fa9c04b3b34a672868accf6a5ed39d9dfeefe97R19), which guards against the USE_DLT define if it detects an Android build. This creates a situation where USE_DLT must be defined for Android, even though Android does not support DLT.
> 
> The introduction of a new define (TRACE_TO_LOGS) with a clear intent is meant to address this technical debt. It decouples the logging functionality from the Android-specific ifdef, making it clear that tracing can be directed to logs across all platforms, not just Android.
> 
> If there’s a way to retain the added functionality while addressing your internal concerns, I’m open to adjusting the implementation. I believe it’s important to resolve this in a way that benefits all platforms without introducing further confusion.

what we were trying to avoid was changes that would involve many files on the repo, because testing and integration of those changes on android side is cumbersome. but from our investigation there is no way around it.

duartenfonseca commented on implementation/tracing/src/connector_impl.cpp:
```diff
@@ -280,9 +273,12 @@ void connector_impl::trace(const byte_t *_header, uint16_t _header_size,
-                std::string app = runtime::get_property("LogApplication");
-
-                ALOGI(app.c_str(), ss.str().c_str());
+                #ifdef ANDROID
+                    std::string app = runtime::get_property("LogApplication");
+                    ALOGI(app.c_str(), ss.str().c_str());
+                #else
```
i however wanted you to try one change, and see if it works on your side:
```suggestion
```
if you remove this part, the VSOMEIP_TRACE should still print on ANDROID, since it has a ifdef guard for android logs

falk-haleytek replied:
Sorry, I'm not sure I follow. Do you mean we should remove that inner ifdef clause where we either log using ALOGI or the VSOMEIP_INFO/TRACE log macro? 

Instead of:

```
#ifdef ANDROID
  std::string app = runtime::get_property("LogApplication");
  ALOGI(app.c_str(), ss.str().c_str());
#else
  VSOMEIP_INFO << ss.str();
#endif
```

Just do:

```
VSOMEIP_INFO << ss.str();
```

BTW, I intentionally opted for using VSOMEIP_INFO here rather than VSOMEIP_TRACE to align with the current master that uses INFO level logging in the Android case (ALOGI).

duartenfonseca replied:
Exactly, if you look into the VSOMEIP_INFO, it already prints for Android:https://github.com/COVESA/vsomeip/blob/637fb6ccce969f89621660dd481badb29a90d661/implementation/logger/src/message.cpp#L144

does this work on your environment?

falk-haleytek replied:
Sure, that would work for us. 

However, I figured that the Android log macros were probably redefined in this file for a reason? Please see current master:  https://github.com/COVESA/vsomeip/blob/master/implementation/tracing/src/connector_impl.cpp#L28

I'm guessing the macros are redefined there to be able to add a specific log tag in this case instead of the default app name. Please see current master: https://github.com/COVESA/vsomeip/blob/master/implementation/tracing/src/connector_impl.cpp#L283

That being said, I don't see the immediate benefit of using a different log tag. So if you prefer, in the interest of keeping this commit as small as possible, I can create another pull request to remove the redefinition AND use of the Android log macro in this file?

duartenfonseca replied:
@falk-haleytek isn't that what also is done here? https://github.com/COVESA/vsomeip/blob/637fb6ccce969f89621660dd481badb29a90d661/implementation/logger/src/message.cpp#L145

falk-haleytek replied:
Sorry, you are completely right! I don't know how I misread that and thought they used different log tags for some reason :) 

I'll update the PR to only do VSOMEIP_INFO << ss.str(); then. 

I guess I should also remove the redefinition of ALOGI in this file then or did I miss the reason for them being there?

duartenfonseca replied:
yes you can remove it!

falk-haleytek replied:
Done! https://github.com/COVESA/vsomeip/pull/726/commits/f37185a2b799e029b0de428aeaad34c2e86a2941

duartenfonseca commented on implementation/tracing/src/connector_impl.cpp:
```diff
@@ -269,7 +247,8 @@ void connector_impl::trace(const byte_t *_header, uint16_t _header_size,
                     VSOMEIP_ERROR << "tracing: found channel without DLT context!";
                 }
-            #else
+            #endif
+            #ifdef TRACE_TO_LOGS
```
last change
```suggestion
            #else

duartenfonseca replied:
this way if you have USE_DLT and TRACE_TO_LOGS defined, you don't print the same message twice

falk-haleytek replied:
No worries, I’m happy to make as many changes as needed to align this PR with the project’s standards. This was a good catch, thank you for pointing it out!

falk-haleytek commented:
@duartenfonseca Just to avoid misunderstandings, are you waiting for anymore input or adjustments from me here?

duartenfonseca commented:
@falk-haleytek we are discussing internally adding this fix to others related to the logger, will reach back to you when that is done

falk-haleytek commented:
@duartenfonseca got it! thanks for the update.

falk-haleytek commented:
@duartenfonseca any updates on the internal discussions?

duartenfonseca commented:
> @duartenfonseca any updates on the internal discussions?

some other issues have appeared, but i will try to get back to this

duartenfonseca commented:
hey @falk-haleytek can you fix the conflicts? i will merge it soon
